[P044] Led settings and events without client, merged into [P020]

tonhuisman opened the pull request:
Resolves #4269 
Resolves #4558

**P044 - Communication - P1 WiFi Gateway**
Features:
- Configure Led pin (Enabled, Pin, Inverted)
- Merged functionality into [P020] Communication - Serial Server to avoid code duplication
  - P044 is still available, is just uses P020 with a predefined, and somewhat limited, configuration options
  - Converts existing [P044] settings to [P020] on first settings-save after update.
- Option to include P1 data in `<taskname>#Data` event (heavy on resources!)

**P020 - Communication - Serial Server**
Features:
- Configure Led pin (Enabled, Pin, Inverted)
- Adds 'P1 WiFi Gateway' option for Event processing
- Increase max. buffer size
- Add option for using Serial Port name as eventname (Generic events only)
- Add option to append the task number to the eventname (Generic and RFLink events only)

**Both plugins**
- Optionally include P1 data in `#data` event (could cause memory overflow issues, especially on ESP8266!), with newlines replaced by commas as P1 data doesn't have any commas
- Select a replacement character for spaces and newlines in the data (before handing over to rules and sending out via TCP)
- Introduce `parseHexTextData()` function in StringConverter
- Add command `serialsendmix,<"text content"|0xXX>[,...]` to concatenate text and hex character values
  - Hex values can be a single byte, or multiple bytes, starting with `0x` and separated by nothing, space, dash or colon.
- Fix plugin initialization when USB (HW)CDC or I2C Serial is selected


TODO:
- [x] Convert settings on first run without prior settings-save
- [x] Documentation P020
  - [x] Document commands
- [x] Documentation P044
- [x] Testing of new P1 WiFi Gateway using the Serial Server code (functionally merged)
- [x] Use buffering supplied by ESPEasySerial

Considering to also add: (feedback requested)
- Fully configurable serial settings _(done)_
- ~Process events without a network client connected ([Forum thread](https://www.letscontrolit.com/forum/viewtopic.php?t=9230))~
- Merge functionality of this plugin into `P020 - Communication - Serial Server`, as suggested [here](https://www.letscontrolit.com/forum/viewtopic.php?t=9230#p59633) _(done)_

tonhuisman committed:
[P044] Uncrustify format source

tonhuisman committed:
[P044] Make Led pin configurable

tonhuisman committed:
[P044] Correct auto-complete typo

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[Plugins] Call PLUGIN_WEBFORM_PRE_SERIAL_PARAMS before serial settings are displayed to allow settings conversion

tonhuisman committed:
[Plugins] Call PLUGIN_WEBFORM_PRE_SERIAL_PARAMS before serial settings are displayed to allow settings conversion

tonhuisman committed:
[P020] Merge [P044] code into [P020], so P020 van emulate P044

tonhuisman committed:
[P020/P044] Migrate settings also on plugin start without prior save from UI

tonhuisman committed:
[P020] Code improvement

tonhuisman committed:
[P020/P044] Add option for including P1 data in #data event

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'feature/P044-led-and-events-without-client' of https://github.com/tonhuisman/ESPEasy-1 into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[UI] Add Separator character input selector

tonhuisman committed:
[P020/P044] Add feature to replace spaces and newlines in received data

tonhuisman committed:
[P020] Add plugin documentation

tonhuisman committed:
[P037] Implement replacement character input selector, update documentation

tonhuisman committed:
[UI] Fix unsupported feature compilation error, fix signed/unsigned warning

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[P044] Only show usable settings

tonhuisman committed:
[P044] Update plugin documentation (based on P020)

TD-er committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

TD-er committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
[Build] Adjust Custom ESP8266 as already planned to fix build failure

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[P020/P044] Adjustments to complete the merge

wdonker commented:
I have tested this PR on my P1-gateway connected to Domoticz. It seems to work ok but in Domoticz log it is showing a warning after every received P1-message:
**Status: P1: WARNING: got new message but buffer still contains unprocessed data from previous message**

This was not the case with the previous implementation. Maybe the plugin is not sending the last character/EOL? I don't know how to see the difference between the messages sent by this and previous implementation.

tonhuisman committed:
[P020] Add extra check before sending P1 data

tonhuisman commented:
@wdonker I've added an extra check to be sure the P1 data is sent out with a terminating `\r\n`, but the code is very similar to the original P044 plugin, as I copied it from there 😉
Can you recheck if this solves the issue?

wdonker commented:
Sorry Ton, now it does not work at all ;-)
Using Telnet I looked at the difference between the original message (220809) and that from version 20130114. I'v pasted them below, you can see that at the end of the P1-message a cr/lf is missing. With the latest version (20230116) the telnet-client does not receive anything. Were there any changes to P044 after 20220809? Then maybe the error was already in the P044-code you used.

Telnet-output of version 20220809:
```
/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(230116102006W)
0-0:96.1.1(4530303039303030303133323439343134)
1-0:1.8.1(010170.010*kWh)
1-0:1.8.2(010578.443*kWh)
1-0:2.8.1(001449.230*kWh)
1-0:2.8.2(003350.765*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(00.405*kW)
1-0:2.7.0(00.000*kW)
0-0:96.7.21(00015)
0-0:96.7.9(00012)
1-0:99.97.0(9)(0-0:96.7.19)(221111154934W)(0000001027*s)(190322092317W)(0000001607*s)(181103094805W)(0000000981*s)(151021144211S)(0000002347*s)(150327105700W)(0000004700*s)(141224155411W)(0000004643*s)(141203111634W)(0000000787*s)(000101000001W)(2147483647*s)(000101000001W)(2147483647*s)
1-0:32.32.0(00000)
1-0:32.36.0(00001)
0-0:96.13.1()
0-0:96.13.0()
1-0:31.7.0(002*A)
1-0:21.7.0(00.403*kW)
1-0:22.7.0(00.000*kW)
0-1:24.1.0(003)
0-1:96.1.0(4730303032333430313536373530333134)
0-1:24.2.1(230116100000W)(07595.970*m3)
!5231
/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(230116102016W)
0-0:96.1.1(4530303039303030303133323439343134)
1-0:1.8.1(010170.010*kWh)
1-0:1.8.2(010578.444*kWh)
1-0:2.8.1(001449.230*kWh)
1-0:2.8.2(003350.765*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(01.046*kW)
1-0:2.7.0(00.000*kW)
0-0:96.7.21(00015)
0-0:96.7.9(00012)
1-0:99.97.0(9)(0-0:96.7.19)(221111154934W)(0000001027*s)(190322092317W)(0000001607*s)(181103094805W)(0000000981*s)(151021144211S)(0000002347*s)(150327105700W)(0000004700*s)(141224155411W)(0000004643*s)(141203111634W)(0000000787*s)(000101000001W)(2147483647*s)(000101000001W)(2147483647*s)
1-0:32.32.0(00000)
1-0:32.36.0(00001)
0-0:96.13.1()
0-0:96.13.0()
1-0:31.7.0(006*A)
1-0:21.7.0(01.046*kW)
1-0:22.7.0(00.000*kW)
0-1:24.1.0(003)
0-1:96.1.0(4730303032333430313536373530333134)
0-1:24.2.1(230116100000W)(07595.970*m3)
!7E1A
```

Output of 20231401:
```
/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(230116102342W)
0-0:96.1.1(4530303039303030303133323439343134)
1-0:1.8.1(010170.010*kWh)
1-0:1.8.2(010578.471*kWh)
1-0:2.8.1(001449.230*kWh)
1-0:2.8.2(003350.765*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(00.410*kW)
1-0:2.7.0(00.000*kW)
0-0:96.7.21(00015)
0-0:96.7.9(00012)
1-0:99.97.0(9)(0-0:96.7.19)(221111154934W)(0000001027*s)(190322092317W)(0000001607*s)(181103094805W)(0000000981*s)(151021144211S)(0000002347*s)(150327105700W)(0000004700*s)(141224155411W)(0000004643*s)(141203111634W)(0000000787*s)(000101000001W)(2147483647*s)(000101000001W)(2147483647*s)
1-0:32.32.0(00000)
1-0:32.36.0(00001)
0-0:96.13.1()
0-0:96.13.0()
1-0:31.7.0(002*A)
1-0:21.7.0(00.410*kW)
1-0:22.7.0(00.000*kW)
0-1:24.1.0(003)
0-1:96.1.0(4730303032333430313536373530333134)
0-1:24.2.1(230116100000W)(07595.970*m3)
!4D89/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(230116102352W)
0-0:96.1.1(4530303039303030303133323439343134)
1-0:1.8.1(010170.010*kWh)
1-0:1.8.2(010578.472*kWh)
1-0:2.8.1(001449.230*kWh)
1-0:2.8.2(003350.765*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(00.406*kW)
1-0:2.7.0(00.000*kW)
0-0:96.7.21(00015)
0-0:96.7.9(00012)
1-0:99.97.0(9)(0-0:96.7.19)(221111154934W)(0000001027*s)(190322092317W)(0000001607*s)(181103094805W)(0000000981*s)(151021144211S)(0000002347*s)(150327105700W)(0000004700*s)(141224155411W)(0000004643*s)(141203111634W)(0000000787*s)(000101000001W)(2147483647*s)(000101000001W)(2147483647*s)
1-0:32.32.0(00000)
1-0:32.36.0(00001)
0-0:96.13.1()
0-0:96.13.0()
1-0:31.7.0(002*A)
1-0:21.7.0(00.409*kW)
1-0:22.7.0(00.000*kW)
0-1:24.1.0(003)
0-1:96.1.0(4730303032333430313536373530333134)
0-1:24.2.1(230116100000W)(07595.970*m3)
!6E04

```

tonhuisman commented:
Thanks, that will help in finding the bug

wdonker commented:
I looked in the original 220809-code of P044. It seems to add cr/lf to the datagram (P044_data_struct.cpp, line 351).

```
  if (done) {
    done = checkDatagram();

    if (done) {
      // add the cr/lf pair to the datagram ahead of reading both
      // from serial as the datagram has already been validated
      addChar('\r');
      addChar('\n');

```

tonhuisman commented:
> It seems to add cr/lf to the datagram

That's still in the code 😃

tonhuisman committed:
[P020] Revert extra check before sending P1 data, minor improvements

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'feature/P044-led-and-events-without-client' of https://github.com/tonhuisman/ESPEasy-1 into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[StringConverter] Add parseHexTextString()

tonhuisman committed:
[P020] Add command `serialsendmix`

TD-er commented on src/src/Helpers/StringConverter.cpp:
```diff
@@ -800,6 +800,47 @@ String tolerantParseStringKeepCase(const String& string, uint8_t indexFind, char
+/*****************************************************************************
+ * handles: 0xXX,text,0xXX," more text ",0xXX starting from index 2 (1-based)
+ ****************************************************************************/
+String parseHexTextString(const String& argument, int index) {
```
Didn't we already have something like this?
After all, the numerical parser does also accept HEX strings.

TD-er commented on src/src/Helpers/StringConverter.cpp:
```diff
@@ -800,6 +800,47 @@ String tolerantParseStringKeepCase(const String& string, uint8_t indexFind, char
+        int hex = -1;
+
+        if (validIntFromString(concat(F("0x"), arg.substring(j, j + 2)), hex) && (hex > -1) && (hex < 256)) {
+          result += char(hex);
```
A string with a `0` char does terminate at that point.
You can only store those in an array (or `std::vector`) of `uint8_t`, not in a String or char array.

tonhuisman replied:
Yes, I realized that too, so the check will be changed to `(hex > 0)` to fix that. No reason to include a 0-char, I think.

tonhuisman committed:
[StringConverter] Improvements to parseHexTextString()

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'feature/P044-led-and-events-without-client' of https://github.com/tonhuisman/ESPEasy-1 into feature/P044-led-and-events-without-client

tonhuisman committed:
[P020] Add command documentation

tonhuisman commented on docs/source/Plugin/P020_commands.repl:
```diff
@@ -0,0 +1,37 @@
+
+        ``<content to send>``: Text that will be sent (nearly) unprocessed. Only the regular variable replacements will be applied before sending the content to the serial port.
+        ","
+        This command will only send data to the serial port if the network client *is* connected.
```
This seems to be misinterpreted (by me), needs some re-writing.

tonhuisman committed:
[P020] Correct some typos in documentation

tonhuisman committed:
[StringConverter] Add parseHexTextData() to handle 0x00

tonhuisman committed:
[P020] Use parseHextTextData() to handle 0x00 in data

tonhuisman committed:
[P020] Update documentation for 0x00..0xFF hex data support in serialsendmix

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[P020] Update changelog

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'feature/P044-led-and-events-without-client' of https://github.com/tonhuisman/ESPEasy-1 into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[P020] Unification of Ser2Net log messages

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
[Build] Manually apply WiFi build fixes (missing includes)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[ESPEasySerial] Add shortName option to be used as eventname

tonhuisman committed:
[Helpers] Implement ESPEasySerial shortName option to be used as eventname

tonhuisman committed:
[P020] Increase max. buffer size
[P020] Add options for using Serial Port name as event and/or append the task number
[P020/P044] Fix plugin initialization when USB (HW)CDC or I2C Serial is selected

tonhuisman committed:
[P020] Update documentation for new options

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'feature/P044-led-and-events-without-client' of https://github.com/tonhuisman/ESPEasy-1 into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[P020] Fix receiving P1 data

tonhuisman commented:
@wdonker Can you retest this build, as I've _finally_ been able to actually test (and fix...) the issues you've mentioned. I'm receiving the P1 data as intended/expected now 😃

wdonker commented:
I will test it next week since I am not at home right now.
Will report back my findings.

TD-er commented on src/src/PluginStructs/P020_data_struct.cpp:
```diff
@@ -464,6 +469,7 @@ bool P020_Task::handleP1Char(char ch) {
     # endif // ifndef BUILD_NO_DEBUG
     state = ParserState::WAITING; // reset
   }
+  ch &= 0x7F;                     // Strip off occasional 8th bit for now
```
Why do this on ser2net?
On the P1 plugin I can understand it, but why on ser2net?

tonhuisman replied:
This method `handleP1Char()` is _only_ called when configured as P1 receiver, and now that the features of P044 (predefined configured for P1 only) have been merged with P020, the code all lives in the P020 sources 😉

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

wdonker commented:
Sorry Ton, the result is still the same: a CR/LF seems to be missing after the last line of the P1 message (I think it is the checksum line).

Output from ESP_Easy_mega_**20230409**_normal_ESP8266_4M1M:

```
/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(230817110236S)
0-0:96.1.1(4530303039303030303133323439343134)
1-0:1.8.1(010739.484*kWh)
1-0:1.8.2(010988.776*kWh)
1-0:2.8.1(001732.340*kWh)
1-0:2.8.2(004047.350*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(00.000*kW)
1-0:2.7.0(00.221*kW)
0-0:96.7.21(00015)
0-0:96.7.9(00012)
1-0:99.97.0(9)(0-0:96.7.19)(221111154934W)(0000001027*s)(190322092317W)(00000016                                                                                                                               07*s)(181103094805W)(0000000981*s)(151021144211S)(0000002347*s)(150327105700W)(0                                                                                                                               000004700*s)(141224155411W)(0000004643*s)(141203111634W)(0000000787*s)(000101000                                                                                                                               001W)(2147483647*s)(000101000001W)(2147483647*s)
1-0:32.32.0(00000)
1-0:32.36.0(00001)
0-0:96.13.1()
0-0:96.13.0()
1-0:31.7.0(001*A)
1-0:21.7.0(00.000*kW)
1-0:22.7.0(00.224*kW)
0-1:24.1.0(003)
0-1:96.1.0(4730303032333430313536373530333134)
0-1:24.2.1(230817100000S)(07995.090*m3)
!9B9C
/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(230817110246S)
etc.
```

Output from ESP_Easy_mega_**20230812**_normal_ESP8266_4M1M:
```
/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(230817110644S)
0-0:96.1.1(4530303039303030303133323439343134)
1-0:1.8.1(010739.484*kWh)
1-0:1.8.2(010988.776*kWh)
1-0:2.8.1(001732.340*kWh)
1-0:2.8.2(004047.368*kWh)
0-0:96.14.0(0002)
1-0:1.7.0(00.000*kW)
1-0:2.7.0(00.350*kW)
0-0:96.7.21(00015)
0-0:96.7.9(00012)
1-0:99.97.0(9)(0-0:96.7.19)(221111154934W)(0000001027*s)(190322092317W)(0000001607*s)(181103094805W)(0000000981*s)(151021144211S)(0000002347*s)(150327105700W)(0000004700*s)(141224155411W)(0000004643*s)(141203111634W)(0000000787*s)(000101000001W)(2147483647*s)(000101000001W)(2147483647*s)
1-0:32.32.0(00000)
1-0:32.36.0(00001)
0-0:96.13.1()
0-0:96.13.0()
1-0:31.7.0(001*A)
1-0:21.7.0(00.000*kW)
1-0:22.7.0(00.350*kW)
0-1:24.1.0(003)
0-1:96.1.0(4730303032333430313536373530333134)
0-1:24.2.1(230817100000S)(07995.090*m3)
!D87C/KFM5KAIFA-METER

1-3:0.2.8(42)
0-0:1.0.0(230817110654S)
etc.


```

tonhuisman commented:
Hm, I'll have a look, as I do have more time available currently, and the hardware to physically test it 😉

tonhuisman committed:
[P020] Fix merge conflict

tonhuisman committed:
[P020] Allow some extra time-out while receiving P1 data, add missing CR/LF

tonhuisman commented:
@wdonker I expect to have found the culprit, please set the RX Receive Timeout to 50-100 msec when using a DSMR 5 meter (115200 baud), so the checksum will be read & calculated in the same 'burst' as the rest of the data. At least my meter needs some time (> 100 msec) to generate the checksum before that's sent.
I've also added the missing CR/LF in the output, as that was read but not added to the buffer before sending it to the client...

wdonker commented:
I think this time you nailed it, Ton!
First I tried this firmware without the RX Receive Timeout but Domoticz complained about a bad CRC.
Than I set the Timeout to 50 mSec and now it works ok. My meter is a DSMR 4 btw.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[P044/P020] Improved defaults for P1 WiFigateway

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[P020][P044] Minor improvements, update with latest mega changes

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P044-led-and-events-without-client

tonhuisman committed:
[P020] Replace WiFiServer->available() by accept() (for IDF >= 5)

TD-er committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client

TD-er committed:
Merge branch 'mega' into feature/P044-led-and-events-without-client
